Added client keepalive arguments to the grpc-proxy

halleyshx opened the pull request:
Added dial-keepalive-time、dial-keepalive-timeout、permit-without-stream arguments to the grpc-proxy for KeepAlive ClientParameters.

Currently cluster's request link: clients(A、B、C) --> grpc-proxy(start by c1.etcd.com) --> etcd-cluster-domain(c1.etcd.com) -->  loadbalancer(11.11.11.11) --> etcd endpoints(e1、e2、e3).

When loadbalancer(11.11.11.11) died, new loadbalancer(11.11.11.12) to replaced, grpc-proxy can't provide service to his clients(A、B、C) for 15min.（why 15min？because gRPC balancergroup's balancerCache is default 15min in gPRC-1.38.0 `var DefaultSubBalancerCloseTimeout = 15 * time.Minute` ), it's too late.

So we need to actively probe for loadbalancer of etcd-cluster-domain survival and reconnect. 
Currently added client keepalive arguments to the grpc-proxy.

This PR introduces three options dial-keepalive-time、dial-keepalive-timeout、permit-without-stream for grpc-proxy：
dial-keepalive-time allows to specify client keepalive interval for grpc-proxy itself (default 0-disable).
dial-keepalive-timeout allows to specify client keepalive timeout for grpc-proxy itself (default 20s).
permit-without-stream allows to specify client sends keepalive pings even with no active RPCs for grpc-proxy itself (default false).

When I start grpc-proxy with `--dial-keepalive-time '20s'`, It worked.

halleyshx commented:
please recheck,thanks @ahrtr

ahrtr commented:
/ok-to-test

halleyshx commented on server/etcdmain/grpc_proxy.go:
```diff
@@ -138,6 +141,9 @@ func newGRPCProxyStartCommand() *cobra.Command {
 	cmd.Flags().DurationVar(&grpcProxyEndpointsAutoSyncInterval, "endpoints-auto-sync-interval", 0, "etcd endpoints auto sync interval (disabled by default)")
+	cmd.Flags().DurationVar(&grpcProxyDialKeepAliveTime, "dial-keepalive-time", 0, "after a duration of this time if the client doesn't see any activity it pings the server to see if the transport is still alive (0 to disable).")
+	cmd.Flags().DurationVar(&grpcProxyDialKeepAliveTimeout, "dial-keepalive-timeout", embed.DefaultGRPCKeepAliveTimeout, "after having pinged for keepalive check, the client waits for a duration of Timeout and if no activity is seen even after that the connection is closed (default 20s).")
+	cmd.Flags().BoolVar(&grpcProxyPermitWithoutStream, "permit-without-stream", false, "if true, client sends keepalive pings even with no active RPCs. if false, when there are no active RPCs, Time and Timeout will be ignored and no keepalive pings will be sent (default false).")
```
Good idea! Please recheck, thanks @ahrtr

ahrtr commented on server/etcdmain/grpc_proxy.go:
```diff
@@ -138,6 +141,9 @@ func newGRPCProxyStartCommand() *cobra.Command {
 	cmd.Flags().DurationVar(&grpcProxyEndpointsAutoSyncInterval, "endpoints-auto-sync-interval", 0, "etcd endpoints auto sync interval (disabled by default)")
+	cmd.Flags().DurationVar(&grpcProxyDialKeepAliveTime, "dial-keepalive-time", 0, "keepalive time for client(grpc-proxy) connections (default 0, disable).")
+	cmd.Flags().DurationVar(&grpcProxyDialKeepAliveTimeout, "dial-keepalive-timeout", embed.DefaultGRPCKeepAliveTimeout, "keepalive timeout for client(grpc-proxy) connections (default 20s).")
+	cmd.Flags().BoolVar(&grpcProxyPermitWithoutStream, "permit-without-stream", false, "if true, client(grpc-proxy) sends keepalive pings even with no active RPCs (default false).")
```
```suggestion
	cmd.Flags().BoolVar(&grpcProxyPermitWithoutStream, "permit-without-stream", false, "Enable client(grpc-proxy) to send keepalive pings even with no active RPCs.")
```

halleyshx replied:
Done.

ahrtr commented:
If you backport the PR to 3.5 (I have no objection on it), then add a changelog for 3.5 after you raise the backporting PR. thx

halleyshx commented:
> If you backport the PR to 3.5 (I have no objection on it), then add a changelog for 3.5 after you raise the backporting PR. thx

OK,I updated the changelog-3.5. Please recheck. thx

ahrtr commented:
> OK,I updated the changelog-3.5. Please recheck. thx

PLease read my previous [comment](https://github.com/etcd-io/etcd/pull/17366#issuecomment-1933807376) . Pls do not update changelog in this PR. Instead, update the changelog after raising the backporting (to 3.5) PR

halleyshx commented:
> > OK,I updated the changelog-3.5. Please recheck. thx
> 
> PLease read my previous [comment](https://github.com/etcd-io/etcd/pull/17366#issuecomment-1933807376) . Pls do not update changelog in this PR. Instead, update the changelog after raising the backporting (to 3.5) PR

Ok. removed. But How should I raise the backporting (to 3.5) PR?

ahrtr commented:
> But How should I raise the backporting (to 3.5) PR?

Make the same change for https://github.com/etcd-io/etcd/tree/release-3.5.

halleyshx commented:
> > But How should I raise the backporting (to 3.5) PR?
> 
> Make the same change for https://github.com/etcd-io/etcd/tree/release-3.5.

Happy new Year!    
Please check https://github.com/etcd-io/etcd/pull/17444  to release-3.5. But where is changelog rewritten, I didn't find it in release-3.5.

ahrtr commented:
> Please check #17444 to release-3.5. But where is changelog rewritten, I didn't find it in release-3.5.

Update https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.5.md#v3513-tbd

ahrtr commented:
> Happy new Year!

thx. Happy new year!

halleyshx commented:
> > Please check #17444 to release-3.5. But where is changelog rewritten, I didn't find it in release-3.5.
> 
> Update https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.5.md#v3513-tbd

Sorry! Please check https://github.com/etcd-io/etcd/pull/17447. 
And Update https://github.com/etcd-io/etcd/blob/main/CHANGELOG/CHANGELOG-3.5.md#v3513-tbd  by this PR? or make a new PR to only update the changelog in CHANGELOG-3.5.md#v3513-tbd？

ahrtr commented:
> make a new PR to only update the changelog in CHANGELOG-3.5.md#v3513-tbd？

Pls create a new PR.

halleyshx commented:
> > make a new PR to only update the changelog in CHANGELOG-3.5.md#v3513-tbd？
> 
> Pls create a new PR.

OK, Pls check https://github.com/etcd-io/etcd/pull/17448 . thx!

halleyshx commented:
> > > make a new PR to only update the changelog in CHANGELOG-3.5.md#v3513-tbd？
> > 
> > 
> > Pls create a new PR.
> 
> OK, Pls check #17448 . thx!

update changelog by https://github.com/etcd-io/etcd/pull/17488
